osu.Game/Overlays/Direct/BeatmapSetPager.cs

```diff
@@ -0,0 +1,87 @@
+using System.Collections.Generic;
+using System.Linq;
+using osu.Game.Beatmaps;
+using osu.Game.Online.API;
+using osu.Game.Online.API.Requests;
+using osu.Game.Rulesets;
+
+namespace osu.Game.Overlays.Direct
+{
+    public class BeatmapSetPager
+    {
+        private readonly IAPIProvider API;
+
+        public event PageFetchHandler PageFetch;
+
+        private readonly RulesetStore rulesets;
+
+        private readonly string query;
+        private readonly RulesetInfo ruleset;
+        private readonly BeatmapSearchCategory searchCategory;
+        private readonly DirectSortCriteria sortCriteria;
+        private readonly SortDirection sortDirection;
+
+        private SearchBeatmapSetsRequest getSetsRequest;
+
+        private int currentPage = 1;
+
+        public bool IsLastPageFetched { get; private set; } = false;
+        public bool IsFetching => getSetsRequest != null;
+        public bool CanFetchNextPage => !IsLastPageFetched && !IsFetching;
+
+        public BeatmapSetPager(IAPIProvider API, RulesetStore rulesets, string query, RulesetInfo ruleset, BeatmapSearchCategory searchCategory = BeatmapSearchCategory.Any, DirectSortCriteria sortCriteria = DirectSortCriteria.Ranked, SortDirection sortDirection = SortDirection.Descending)
+        {
+            this.API = API;
+
+            this.rulesets = rulesets;
+
+            this.query = query;
+            this.ruleset = ruleset;
+            this.searchCategory = searchCategory;
+            this.sortCriteria = sortCriteria;
+            this.sortDirection = sortDirection;
+        }
+
+        public void FetchNextPage()
+        {
+            if (getSetsRequest != null)
+                return;
+
+            getSetsRequest = new SearchBeatmapSetsRequest(
+                query,
+                ruleset,
+                currentPage,
+                searchCategory,
+                sortCriteria,
+                sortDirection);
+
+            lock (getSetsRequest) {
+                getSetsRequest.Success += response =>
+                {
+                    var sets = response.BeatmapSets.Select(r => r.ToBeatmapSet(rulesets)).ToList();
+
+                    if (sets.Count <= 0) IsLastPageFetched = true;
+
+                    PageFetch?.Invoke(currentPage, sets);
+
+                    getSetsRequest = null;
+                    currentPage++;
+                };
+
+                API.Queue(getSetsRequest);
+            }
```

Comment:
Not sure if this is correct. Can't lock on the request before this section because it may be null.

Reply:
I suppose I could also make a dummy object and use that to lock the whole method if desired.

Reply:
This locking is definitely not correct. This doesn't really do what it appears to be doing (it takes the lock for the duration of subscribing to the event and queueing a request, and actually doesn't keep hold of it *inside* the response callback), and it doesn't prevent the race either. The actual race is slightly higher, in between the read of `getSetsRequest` and assignment to it after the null-check.

But honestly now that I look at this code again after a while the chance of this racing is probably so slim that it might not really be worth checking for. Probably fine to ignore that but mention in the xmldoc that the method is not thread-safe so that users tread carefully.

Reply:
I think that it's quite unlikely for that to happen as well. I'll just remove the lock and add a comment like you suggested.